### PR TITLE
Add forgotten status to Room Details API

### PR DESCRIPTION
**Pull request by dklimpel**
Relates to:
- #12170

Add requested tests:
- https://github.com/matrix-org/synapse/pull/12170#pullrequestreview-904151297

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: tests/storage/test_roommember.py [hunk 1]
```diff
@@ -157,6 +157,55 @@ def test__null_byte_in_display_name_properly_handled(self) -> None:
...
+
+        # local users leave the room and the room is not forgotten
+        self.inject_room_member(self.room, self.u_alice, Membership.LEAVE)
+        self.inject_room_member(self.room, self.u_bob, Membership.LEAVE)
+        self.assertFalse(
+            self.get_success(self.store.is_locally_forgotten_room(self.room))
+        )
+
+        # first user forgets the room, room is not forgotten
+        self.assertNoResult(self.store.forget(self.u_alice, self.room))
```
Comment (dklimpel) [inline tests/storage/test_roommember.py:178]: I am need a hint for the failing test:
```
Traceback (most recent call last):
  File "/github/workspace/tests/storage/test_roommember.py", line 178, in test_room_is_locally_forgotten
    self.assertNoResult(self.store.forget(self.u_alice, self.room))
  File "/github/home/.cache/pypoetry/virtualenvs/matrix-synapse-cDQ_M3T8-py3.8/lib/python3.8/site-packages/twisted/trial/_synctest.py", line 777, in assertNoResult
    deferred.addBoth(cb)
builtins.AttributeError: 'coroutine' object has no attribute 'addBoth'

tests.storage.test_roommember.RoomMemberStoreTestCase.test_room_is_locally_forgotten
```

Is the solution?
```python
        forget = ensureDeferred(self.store.forget(self.u_alice, self.room))
        self.assertNoResult(forget)
```

Comment (squahtx) [reply tests/storage/test_roommember.py:178]: `assertNoResult` asserts that a deferred hasn't completed yet. `get_success` and ignoring the result would be more appropriate:

```suggestion
        self.get_success(self.store.forget(self.u_alice, self.room))
```

File: synapse/storage/databases/main/roommember.py [hunk 2]
```diff
@@ -1212,6 +1212,30 @@ def _get_forgotten_rooms_for_user_txn(txn: LoggingTransaction) -> Set[str]:
...
+        Returns:
+            Whether the room is forgotten.
+        """
+
+        sql = """
+            SELECT count(*) FROM local_current_membership
+            INNER JOIN room_memberships USING (room_id, event_id)
+            WHERE
+                room_id = ?
+                AND forgotten = 0;
```
Comment (DMRobertson) [inline synapse/storage/databases/main/roommember.py:1230]: Drive-by comment: it might(?) be more performant to `SELECT count(*) > 0` here. That gives the database a chance to stop the query as soon as it finds the first membership that hasn't been forgotten.

(There's also an index `"room_memberships_user_room_forgotten" btree (user_id, room_id) WHERE forgotten = 1` but I'm not sure if there's an easy way to use that here.)

Comment (dklimpel) [reply synapse/storage/databases/main/roommember.py]: https://github.com/matrix-org/synapse/blob/8fa7eb8bcc31bb46bbf3396884fdd97d56160259/synapse/storage/databases/main/roommember.py#L1188-L1207
This looks similar, but is very complicated.

Comment (dklimpel) [reply synapse/storage/databases/main/roommember.py]: I try to create a better sql query. I am feeling in a rabbit hole. It seems that Synapse is not possible to forget a room, at the moment. I think there is a bug in:
https://github.com/matrix-org/synapse/blob/e825f7366b589f61977f63d4fdc598abe052e6a0/synapse/handlers/room_member.py#L1912-L1927
There is a gap of unit tests.

Comment (dklimpel) [reply synapse/storage/databases/main/roommember.py]: Proposal for an other query:
```sql
SELECT room_id, ( 
 SELECT count(*) > 0 FROM room_memberships 
 WHERE room_id = m.room_id AND forgotten = 0 
) AS count 
FROM room_memberships AS m 
WHERE room_id = ? AND forgotten = 1 
GROUP BY room_id;
```

Steps
1. Query if the room is at least one time forgotten - uses partial index on `forgotten = 1`
2. Query this rooms (result of no. 1) that there is no user which has not forgotten (or rejoin) this room

Results
- `count == 0` -> room is forgotten
- `count > 0` -> room is not fully forgotten but at least one time
- no result / `count is None` -> room is not in list (unknown) or room is not at least one time forgotten

File: synapse/storage/databases/main/roommember.py [hunk 3]
```diff
@@ -1215,6 +1215,30 @@ def _get_forgotten_rooms_for_user_txn(txn: LoggingTransaction) -> Set[str]:
...
+            Whether the room is forgotten.
+        """
+
+        sql = """
+            SELECT count(*) > 0 FROM local_current_membership
+            INNER JOIN room_memberships USING (room_id, event_id)
+            WHERE
+                room_id = ?
+                AND forgotten = 0;
+        """
```
Comment (erikjohnston) [inline synapse/storage/databases/main/roommember.py:1234]: The query plan looks sensible to me

**Review by erikjohnston — APPROVED** [review]

File: docs/admin_api/rooms.md [hunk 4]
```diff
@@ -302,6 +302,8 @@ The following fields are possible in the JSON response body:
 * `state_events` - Total number of state_events of a room. Complexity of the room.
 * `room_type` - The type of the room taken from the room's creation event; for example "m.space" if the room is a space.
   If the room does not define a type, the value will be `null`.
+* `forgotten` - Whether all local users have
+  [forgotten](https://spec.matrix.org/latest/client-server-api/#leaving-rooms) the room.
```
Comment (DMRobertson) [inline docs/admin_api/rooms.md:306]: I think we should note that this was added in Synapse 1.66. I'll add something to this file now while preparing the release.
